File: ch04/20/printMatrix.cpp
class Solution {
public:
    vector<int> printMatrix(vector<vector<int> > matrix) {
        vector<int> re_vec;
        int rows = matrix.size();
        int cols = matrix[0].size();
        if (0 == rows || 0 == cols)
            return re_vec;

        int start = 0;
        while(rows > 2 * start && cols > 2 * start){
            vector<int> tmp_vec = getNumberInCircle(matrix, rows, cols, start);
            re_vec.insert(re_vec.end(), tmp_vec.begin(), tmp_vec.end());
            start ++;
        }
        return re_vec;
    }

    vector<int> getNumberInCircle(vector<vector<int> > matrix, int rows, int cols, int start) {
        vector<int> re;
        int endX = cols - start - 1; // 每一圈结束的行列坐标
        int endY = rows - start - 1; 

        for (int i = start; i <= endX; i ++)
            re.push_back(matrix[start][i]);
        
        // 从上往下
        if (endY > start) {
            for (int i = start + 1; i <= endY; i++)
                re.push_back(matrix[i][endX]);
        }

        // 从右往左
        if (endX > start && endY > start) {
            for (int i = endX - 1; i >= start; --i)
                re.push_back(matrix[endY][i]);
        }

        // 从下往上
        if (endY > start + 1 && endX > start) {
            for (int i = endY - 1; i > start; --i)
                re.push_back(matrix[i][start]);
        }

        return re;
    }
};

File: ch04/20/printMatrix.py
# -*- coding:utf-8 -*-
# 输入一个矩阵，按照从外向里以顺时针的顺序依次打印出每一个数字
class Solution:
    # matrix类型为二维列表，需要返回列表
    def printMatrix(self, matrix):
        # write code here
        re_list = []
        rows = len(matrix)
        cols = len(matrix[0])
        if 0 == rows or 0 == cols:
            return re_list
        start = 0

        while rows > start * 2 and cols > start * 2:
            re_list.extend(self.getNumbersInCircle(matrix, rows, cols, start))
            start += 1
        return re_list

    def getNumbersInCircle(self, matrix, rows, cols, start):
        re = []
        sub_rows = rows - 2 * start # 需要经过的行数
        sub_cols  = cols - 2 * start
        
        # 从左往右
        for i in range(sub_cols):
            re.append(matrix[start][start + i])
        
        # 从上往下
        if sub_rows >=2:
            for i in range(1, sub_rows):
                re.append(matrix[start + i][start + sub_cols - 1])
        
        # 从右往左
        if sub_rows >= 2 and  sub_cols >= 2:
            for i in range(start + sub_cols - 2, start - 1, -1):
                re.append(matrix[start + sub_rows - 1][i])

        # 从下往上
        if sub_rows > 2 and sub_cols >=2:
            for i in range(start + sub_rows - 2, start, -1):
                re.append(matrix[i][start])
        
        return re

if __name__ == '__main__':
    m = [[1,2,3,4], [5,6,7,8], [9,10,11,12], [13,14,15,16]]
    m2 = [[1],[2],[3],[4],[5]]
    s = Solution()
    print s.printMatrix(m2) 

